src/reduce_dims.cpp

```diff
@@ -0,0 +1,116 @@
+#include <migraphx/reduce_dims.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+bool reduce_dim(std::vector<shape>& shapes, std::size_t n)
+{
+    std::vector<std::size_t> new_lens;
+    for(const auto& s : shapes)
+    {
+        assert(n < s.lens().size());
+        if((n + 1) >= s.lens().size())
+            return false;
+        auto astride = s.strides()[n];
+        auto alen    = s.lens()[n];
+        auto bstride = s.strides()[n + 1];
+        auto blen    = s.lens()[n + 1];
+
+        if(astride == bstride * blen)
+        {
+            new_lens.push_back(alen * blen);
+        }
+    }
+    if(new_lens.size() != shapes.size())
+        return false;
+    std::size_t i = 0;
+    for(auto& s : shapes)
+    {
+        auto lens    = s.lens();
+        auto strides = s.strides();
+        lens.erase(lens.begin() + n);
+        strides.erase(strides.begin() + n);
+        lens[n] = new_lens[i];
+        s       = shape{s.type(), lens, strides};
+        i++;
+    }
+    return true;
+}
+
+std::size_t reduce_dim_all(std::vector<shape>& shapes, std::size_t n)
+{
+    while(reduce_dim(shapes, n))
+    {
```

Comment:
It will return `false` when it can no longer collapse the dimensions at `n`.

Reply:
there is no change about n if it returns true, so in this case it going to be an infinite loop. You need a ++n in side the loop?

Reply:
No it wont be an infinite loop. It returns `true` when it can collapse the dimension. For example, if we have dimensions `x,y,z`. It will first collapse the `x,y` so this will become `xy,z`. Then it will then start at `n==0` again and it will collapse `xy,z` so it will become `xyz`. After this it can no longer collapse the dimensions, so it will return `false`.